include/proxy-wasm/wasm_vm.h

```diff
@@ -303,6 +303,12 @@ class WasmVm {
    */
   virtual void terminate() = 0;
 
+  /**
+   * Byte order flag (host or wasm).
+   * @return 'false' for a null VM and 'true' for a wasm VM.
+   */
+  virtual bool isWasmByteOrder() = 0;
```

Comment:
I have a question/comment, but it runs contrary to the suggestions provided thus far for this PR by people more knowledgeable about the code than I, so take with a grain of salt:

I don't fully understand why this adjustment is being made in generic host code as opposed to fully within NullVm. In other words, why not state that host <-> VM values are always passed as Wasm byte order, and if the VM happens to use the same endianness as the host, it is that specific VM implementation's responsibility to unswap the bytes back from Wasm byte order to its native order? IMO having a consistent byte-order at the host <-> VM boundary is appealing in that it is simple and easy to reason about.

Is the concern that swapping + unswapping is too much of a performance hit? I guess I'd hope that a compiler could optimize that away, though maybe the virtual dispatch confounds that.

Reply:
In case of wasm VMs that use wasm binary (which is LE reinforced) each wasm engine (v8, wasmtime, ...) is responsible for reversing bytes to LE when running on BE machine (and therefore we have to take it into the account also in wasm-proxy marshalPairs/toPairs and reverse back on BE, so we are already swapping bytes on BE in that generic host code to fix sizes/pointers).
In case of null VM we don't have wasm binary / wasm engine: nullvm gets compiled natively into the different C++ code (envoy, istio proxy, ...) and there is no a dedicated null vm engine, that makes bytecode LE reinforced.  So it is simpler to fix proxy-wasm than changing all other repos that are compiled with wasm-proxy.
Another reason to make a fix for null-vm use case in proxy-wasm (and not in other repos) is that null-vm scenario has always worked using host native byte order, and the problem appeared only when we introduced BE support in proxy-wasm https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/198 and by mistake started reversing bytes for all wasm VMs including nullvm. So I think it is better to fix that regression in proxy-wasm than changing rules for any external code that links with proxy-wasm and uses nullvm.

Reply:
Thanks for the explanation, SGTM